File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
N, K = map(int, input().split())
array = [0]
for _ in range(N):
    array.append(list(map(int, input().split())))

dp = [[0] * (K + 1) for _ in range(N + 1)]

# 무게에 대해 물건이 포함 되는지 안되는지 확인
for i in range(1, N + 1):
    w, v = array[i]

    for kg in range(K + 1):
        if kg - w >= 0:
            dp[i][kg] = max(dp[i-1][kg], dp[i-1][kg-w] + v)
        else:
            dp[i][kg] = dp[i-1][kg]

print(dp[-1][-1])

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.swift
var temp = readLine()!.split(separator: " ").map{ Int($0)! }
var N = temp[0]
var K = temp[1]

var items = [(Int, Int)]()
items.append((0, 0))
for _ in 1...N {
    let temp = readLine()!.split(separator: " ").map{ Int($0)! }
    items.append((temp[0], temp[1]))
}

// col - 무게, row - 물건별
var dp = Array(repeating: [Int](repeating: 0, count: K + 1), count: N + 1)

for i in 1...N {
    let (W, V) = items[i]
    for k in 1...K {
        if k - W >= 0 {
            dp[i][k] = max(dp[i - 1][k - W] + V, dp[i - 1][k])
        } else {
            dp[i][k] = max(dp[i][k - 1], dp[i - 1][k])
        }
       
    }
}

print(dp[N].last!)
